### PR TITLE
[IMP] account_check_printing: speed up dashboard w/ index

### PR DESCRIPTION
**Pull request by moylop260**
Install "account_check_printing" module
Open "Accounting" icon from the main menu
It opens the journal dashboard by default

It runs the following search domain in account.payment model:
```python
    domain = [
        ...
        ('payment_method_id.code', '=', 'check_printing'),
        ...
    ]
```

It runs the following query:

```sql
    SELECT "account_payment".id
    FROM "account_payment"
    WHERE ...
        AND  ("account_payment"."payment_method_id" in ($2)))
        ...
```

The average duration of this query is 37ms

It query is ran for each journal created
If you have created 600 journals (real case) so it will run 600 times
It will spend more than 22 seconds opening this dashboard
After that, you are available to create accounting actions since that
the menu is not available before of this dashboard
So, it is important to open it faster

Creating the following index:
    - `account_payment(payment_method_id)`

The average duration of the last query is reduced to 0.935ms instead.
It means, it will spend 0.5s opening this important dashboard with 600 journals

44x faster

<details><summary>SQL Analyze Results</summary>
<p>

# Explain analyze without index
```sql
EXPLAIN (ANALYZE, VERBOSE, BUFFERS) SELECT id FROM "account_payment" WHERE ((("account_payment"."journal_id" = 1245) AND ("account_payment"."payment_method_id" IN (3))) AND ("account_payment"."state" = 'posted')) AND (FALSE OR ("account_payment"."journal_id" IN (11, 12, 13, 16, ...)))
```

Result:
```txt
+------------------------------------------------------------------------------------------------------------------------------------------------------
| QUERY PLAN
|------------------------------------------------------------------------------------------------------------------------------------------------------
| Gather  (cost=1000.00..49242.18 rows=1 width=4) (actual time=8.115..10.151 rows=0 loops=1)
|   Output: id
|   Workers Planned: 2
|   Workers Launched: 2
|   Buffers: shared hit=3862
|   ->  Parallel Seq Scan on public.account_payment  (cost=0.00..48242.08 rows=1 width=4) (actual time=5.835..5.835 rows=0 loops=3)
|         Output: id
|         Filter: ((account_payment.journal_id = 1245) AND (account_payment.payment_method_id = 3) AND ((account_payment.state)::text = 'posted'::text)
|         Rows Removed by Filter: 38855
|         Buffers: shared hit=3862
|         Worker 0: actual time=4.881..4.882 rows=0 loops=1
|           Buffers: shared hit=994
|         Worker 1: actual time=5.020..5.020 rows=0 loops=1
|           Buffers: shared hit=1025
| Planning time: 0.274 ms
| Execution time: 10.164 ms
+------------------------------------------------------------------------------------------------------------------------------------------------------
```

# Creating index only using `payment_method_id`
 - `CREATE INDEX x ON account_payment (payment_method_id);`

```txt
+------------------------------------------------------------------------------------------------------------------------------------------------------
| QUERY PLAN
|------------------------------------------------------------------------------------------------------------------------------------------------------
| Aggregate  (cost=2106.69..2106.70 rows=1 width=8) (actual time=0.192..0.192 rows=1 loops=1)
|   Output: count(1)
|   Buffers: shared hit=174
|   ->  Bitmap Heap Scan on public.account_payment  (cost=12.79..2106.68 rows=1 width=0) (actual time=0.190..0.190 rows=0 loops=1)
|         Output: id, message_main_attachment_id, name, state, payment_type, payment_reference, move_name, destination_journal_id, multi, payment_metho
|         Recheck Cond: (account_payment.payment_method_id = 3)
|         Filter: ((account_payment.journal_id = 1245) AND ((account_payment.state)::text = 'posted'::text) AND (account_payment.journal_id = ANY ('{11
|         Rows Removed by Filter: 179
|         Heap Blocks: exact=170
|         Buffers: shared hit=174
|         ->  Bitmap Index Scan on x  (cost=0.00..12.79 rows=583 width=0) (actual time=0.018..0.018 rows=179 loops=1)
|               Index Cond: (account_payment.payment_method_id = 3)
|               Buffers: shared hit=4
| Planning time: 0.293 ms
| Execution time: 0.216 ms
+------------------------------------------------------------------------------------------------------------------------------------------------------
```

# Creating indexes using `payment_method_id` and `state` separated:
 - `CREATE INDEX x ON account_payment (payment_method_id);`
 - `CREATE INDEX y ON account_payment (state);`

```txt
+------------------------------------------------------------------------------------------------------------------------------------------------------
| QUERY PLAN
|------------------------------------------------------------------------------------------------------------------------------------------------------
| Aggregate  (cost=44.32..44.33 rows=1 width=8) (actual time=1.431..1.431 rows=1 loops=1)
|   Output: count(1)
|   Buffers: shared hit=17 read=53
|   ->  Bitmap Heap Scan on public.account_payment  (cost=29.83..44.32 rows=1 width=0) (actual time=1.429..1.429 rows=0 loops=1)
|         Output: id, message_main_attachment_id, name, state, payment_type, payment_reference, move_name, destination_journal_id, multi, payment_metho
|         Recheck Cond: ((account_payment.payment_method_id = 3) AND ((account_payment.state)::text = 'posted'::text))
|         Filter: ((account_payment.journal_id = 1245) AND (account_payment.journal_id = ANY ('{11,12,13,16,548,14,1245,1246,1247,1153,1060,81,68,80,69
|         Rows Removed by Filter: 18
|         Heap Blocks: exact=17
|         Buffers: shared hit=17 read=53
|         ->  BitmapAnd  (cost=29.83..29.83 rows=3 width=0) (actual time=1.408..1.408 rows=0 loops=1)
|               Buffers: shared read=53
|               ->  Bitmap Index Scan on x  (cost=0.00..12.79 rows=583 width=0) (actual time=0.053..0.053 rows=179 loops=1)
|                     Index Cond: (account_payment.payment_method_id = 3)
|                     Buffers: shared read=4
|               ->  Bitmap Index Scan on y  (cost=0.00..16.79 rows=583 width=0) (actual time=1.344..1.344 rows=16819 loops=1)
|                     Index Cond: ((account_payment.state)::text = 'posted'::text)
|                     Buffers: shared read=49
| Planning time: 0.443 ms
| Execution time: 1.465 ms
+------------------------------------------------------------------------------------------------------------------------------------------------------
```

# Creating complex index using `payment_method_id` and `state` together:
 - `CREATE INDEX y ON account_payment (payment_method_id, state);`

```txt
+------------------------------------------------------------------------------------------------------------------------------------------------------
| QUERY PLAN
|------------------------------------------------------------------------------------------------------------------------------------------------------
| Aggregate  (cost=18.94..18.95 rows=1 width=8) (actual time=0.036..0.036 rows=1 loops=1)
|   Output: count(1)
|   Buffers: shared hit=21
|   ->  Bitmap Heap Scan on public.account_payment  (cost=4.45..18.94 rows=1 width=0) (actual time=0.035..0.035 rows=0 loops=1)
|         Output: id, message_main_attachment_id, name, state, payment_type, payment_reference, move_name, destination_journal_id, multi, payment_metho
|         Recheck Cond: ((account_payment.payment_method_id = 3) AND ((account_payment.state)::text = 'posted'::text))
|         Filter: ((account_payment.journal_id = 1245) AND (account_payment.journal_id = ANY ('{11,12,13,16,548,14,1245,1246,1247,1153,1060,81,68,80,69
|         Rows Removed by Filter: 18
|         Heap Blocks: exact=17
|         Buffers: shared hit=21
|         ->  Bitmap Index Scan on y  (cost=0.00..4.45 rows=3 width=0) (actual time=0.013..0.013 rows=18 loops=1)
|               Index Cond: ((account_payment.payment_method_id = 3) AND ((account_payment.state)::text = 'posted'::text))
|               Buffers: shared hit=4
| Planning time: 0.308 ms
| Execution time: 0.061 ms
+------------------------------------------------------------------------------------------------------------------------------------------------------
```

The best time was using the complex index.

UPDATE: Master version is not using this field anymore.

</p>
</details>

### CODE REVIEW

**Comment by moylop260** [comment]
@odony 
What do you think about this index?

File: addons/account_check_printing/models/account_payment.py [hunk 1]
```diff
@@ -263,3 +263,9 @@ def _check_make_stub_line(self, invoice):
             'amount_paid': formatLang(self.env, invoice_sign * amount_paid, currency_obj=self.currency_id),
             'currency': invoice.currency_id,
         }
+
+    def _auto_init(self):
```
Comment (moylop260) [inline addons/account_check_printing/models/account_payment.py:267]: Should we use `def _auto_init` or `def init`?

I ran the following rgrep command to detect the most common method used:
 - `rgrep "tools.create_index" . --include=*.py -B3 |grep def`

And it is `def _auto_init` but it is private (_) and the docstring looks like best option `def init`
For that, I'm confused what should we use here

**Comment by odony** [comment]
@moylop260 Thanks for the analysis. I think `account_payment` doesn't have a `state` field anymore, it is inherited from `account.move` in master, and that won't be indexable.

Further, considering that most payments end up `posted`, I'm not sure that composite index on `state` brought any significant speed gain to begin with.

You could probably just set `index=True` on `payment_method_id`, for almost the same gain, and the patch will be better:
- much less code
- less specific index, otherwise it would be likely to become useless after some dashboard refactoring and be left to rot

PS: if would be nice to adapt the commit message a bit, e.g. it should be `[IMP]` and with a short headline, like:
`[IMP] account_check_printing: speed up dashboard w/ index`.
The rationale in the body of the commit message is great :+1:

**Comment by moylop260** [comment]
Hi @odony 
Thanks for reply

> I think account_payment doesn't have a state field anymore

Checking the runbot message error looks like that.
I'm curious the original domain has `state` field yet in master branch:
 - https://github.com/odoo/odoo/blob/d7af73dcb1b4ef3ada39bcafa7ad23972e487b45/addons/account_check_printing/models/account_journal.py#L115

(But it is another matter)

I have changed the commit message and I have created the index directly re-defining the field.
Thank you!

**Review by odony — APPROVED** [review]
@moylop260 works for me, thanks!

@robodoo r+

/cc @qdp-odoo in case you want to add indexes elsewhere for similar purposes

**Pull request closed without merging**